src/main/scala/z3/scala/dsl/ValHandler.scala

```diff
@@ -20,9 +20,9 @@ abstract class ValHandler[A : Default] {
   /** Z3 code to construct a sort representing the Scala A type. */
   def mkSort(z3 : Z3Context) : Z3Sort
 
-  /** Constructs a Val[A], i.e. the representation of a variable of type A. */
-  def construct : Val[A] = new Val[A] {
-    def build(z3 : Z3Context) = z3.mkFreshConst("valCst", mkSort(z3))
+  /** Constructs a ValTree[A], i.e. the representation of a variable of type A. */
+  def construct[T >: BottomSort <: TopSort]: Tree[T] = new ValTree[T] {
```

Comment:
Also, if I am not wrong, having a second type parameter to `ValHandler` requires to change all its uses as typeclass (eg. in type parameter `A : ValHandler`) to explicitly use implicit parameters.